docs/proposals/scalable-query-frontend.md

```diff
@@ -0,0 +1,267 @@
+  desiredConcurrency int = 3;
+}
+```
+
+**currentConcurrency** - The current number of queries being executed by the querier.
+
+**desiredConcurrency** - The total number of queries that a querier is capable of executing.
+
+Add a short backoff to the main frontend [processing loop](https://github.com/cortexproject/cortex/blob/8fb86155a7c7c155b8c4d31b91b267f9631b60ba/pkg/querier/frontend/frontend.go#L288-L331).  This would cause the frontend to briefly back off of any querier that was overloaded but continue to send queries to those that were capable of doing work.
+
+```go
+if current > desired {
+  zzz := (current - desired) * backoffDuration
+  zzz *= 1 + rand.Float64() * .1               // jitter
+  time.Sleep(zzz)
+}
+```
+
+Passing flow control information from the querier to the frontend would also open up additional future work for more sophisticated load balancing across queriers.  For example by simply comparing and choosing [the least congested of two](https://www.nginx.com/blog/nginx-power-of-two-choices-load-balancing-algorithm/) queriers we could dramatically improve how well work is distributed.
+
+### Increased Time To Failure
+
+#### Challenge
+
+Scaling the query frontend also increases the per tenant queue length by creating more queues.  This could result in increased latencies where failing fast (429) would have been preferred.
+
+The operator could reduce the queue length per query frontend in response to scaling out, but
+
+then they would run the risk of unnecessarily failing a request due to unbalanced distribution across query frontends.  Also, shorter queues run the risk of failing to properly service heavily sharded queries.
+
+Another concern is that a system with more queues will take longer to recover from an production event as it will have queued up more work.
+
+#### Proposal
+
+Currently we are not proposing any changes to alleviate this concern.  We believe this is solvable operationally.  This can be revisited as more information is gathered.
+
+### Querier Discovery Lag
+
+#### Challenge
+
+Queriers have a configurable parameter that controls how often they refresh their query frontend list.  The default value is 10 seconds.  After a new query frontend is added the average querier will take 5 seconds (after DNS is updated) to become aware of it and begin requesting queries from it.
+
+#### Proposal
+
+It is recommended to add a readiness/health check to the query frontend to prevent it from receiving queries while it is waiting for queriers to connect.   HTTP health checks are supported by [envoy](https://www.envoyproxy.io/learn/health-check), [k8s](https://kubernetes.io/docs/tasks/configure-pod-container/configure-liveness-readiness-startup-probes/), [nginx](https://docs.nginx.com/nginx/admin-guide/load-balancer/http-health-check/), and basically any commodity load balancer.  The query frontend would not indicate healthy on its health check until at least one querier had connected.
```

Comment:
No this is a good point.  k8s readiness is not an appropriate health check for the reasons you outlined.  I conflated the ideas of service discovery and load balancing here.

Removing it from the list.

Reply:
I have added a paragraph describing the problems with a k8s deployment and the proposed healthcheck.  For this reason (and to distinguish general health from query-frontend health) I believe we should make a new health check endpoint explicitly for this.

Reply:
I'm not sure I can understand your proposal. You propose to add a new dedicated health check endpoint, but who calls this health check endpoint?

Reply:
The load balancer would use the new healthcheck to determine if it should send traffic to a given frontend.  It could even be the existing endpoint with a simple param.  Something like:
```
/ready?require-queriers=true
```
This would allow k8s deployments to continue being configured as they are, but also allow a load balancer to only send traffic to those query frontends that had an attached querier

Reply:
Apologies, disregard my last comment.  There is a feature of k8s services for this exact scenario:  `publishNotReadyAddresses`

This will allow a headless service to be used as discovery for queriers.

Reply:
Thanks for addressing my feedback. Even if it's out of the scope of this design (because it's just an implementation detail) I would suggest to use `/query-frontend/ready` instead of a query string parameter like `/ready?require-queriers=true` because `/ready` is a generic handler and would require some hacks to plug into some logic specific to a single component.